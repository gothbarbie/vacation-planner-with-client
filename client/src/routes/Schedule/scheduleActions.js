import axios from 'axios'

import { CREATE_VACATION, FLASH_MESSAGE } from '../../actions/types'

export const createVacation = ({
  arrival,
  departure,
  people,
}) => async dispatch => {
  try {
    const res = await axios.post('/api/vacation/create', {
      arrival,
      departure,
      people,
    })
    dispatch({
      type: CREATE_VACATION,
      payload: res.data,
    })
  } catch (error) {
    const payload = {
      type: 'danger',
      message: 'Could not save new trip.',
    }
    dispatch({
      type: FLASH_MESSAGE,
      payload,
    })
  }
}
