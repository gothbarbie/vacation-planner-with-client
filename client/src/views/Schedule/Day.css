.day {
  background: #F2F2F2;
  font-family: 'Avenir Next', sans-serif;
  font-weight: 600;
  font-size: 18px;
  color: #828282;
  border: 1px solid #F2F2F2;
}

.day:hover {
  border: 1px solid #80D1FF;
}

.day--weekend {
  color: #EB5757;
}

.day--empty {
  background: #fff;
}

.day--occupied {
  border: 1px solid #EB5757;
}

.day__date {
  margin: .75rem;
}

.day__people {
  display: flex;
  justify-content: center;
  align-items: center;
  min-height: 3.5rem;
}

.day__status {
  font-family: 'Avenir Next', sans-serif;
  font-size: 13px;
  font-weight: 600;
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  height: 26px;
}

.day__status--occupied {
  color: #FFFFFF;
  background: #EB5757;
}
