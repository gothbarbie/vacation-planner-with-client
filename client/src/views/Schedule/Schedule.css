.schedule {
  width: 100%;
}

.schedule__inner {
  display: grid;
  grid-template-columns: 60% auto 30%;
  grid-template-rows: auto;
  justify-content: space-between;
  align-items: center;
}

.calendar{
  background: #FAFAFA;
  width: 100%;
  margin-top: 5rem;
}

.calendar__header {
  font-family: 'Avenir Next', sans-serif;
  color: #828282;
  font-weight: 600;
  font-size: 18px;
  height: 3rem;
  display: grid;
  grid-template-columns: repeat(7, 1fr);
  grid-column-gap: .75rem;
  justify-items: center;
  align-items: center;
  border-bottom: 2px solid #80D1FF;
  padding: 0 1.75rem;
}

.calendar__days {
  display: grid;
  grid-template-columns: repeat(7, 1fr);
  grid-gap: .75rem .75rem;
  padding: 1.75rem;
}

.time-nav {
  display: flex;
  align-items: center;
  justify-content: space-evenly;
  align-items: center;
  margin-top: 2rem;
}

.time-nav a {
  color: #000;
  display: flex;
  align-items: center;
  font-family: 'Avenir Next', sans-serif;
  text-decoration: none;
}

.time-nav a:hover {
  color: #80D1FF;
  text-decoration: underline;
}
