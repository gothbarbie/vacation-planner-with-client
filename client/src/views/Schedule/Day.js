// @flow

import React from 'react'
import classnames from 'classnames'

import './Day.css'

type props = {
  date?: number,
  occupied: boolean,
  people: [{
    name: string
  }],
  status?: 'string',
  weekend: boolean,
  empty: boolean,
}

export default ({ date, empty, occupied, people, status, weekend }: props) => (
  <div
    className={classnames('day', {
      'day--weekend': weekend,
      'day--occupied': occupied,
      'day--empty': empty,
    })}>
    <div className="day__date">{date}</div>
    <div className="day__people">
      {people &&
        people.map(person => {
          return <div>{person.name}</div>
        })}
    </div>
    <div className={classnames('day__status', {
      'day__status--occupied': occupied,
    })}>{status}</div>
  </div>
)
