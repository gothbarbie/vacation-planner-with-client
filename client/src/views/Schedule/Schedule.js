// @flow

import React, { Component } from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import moment from 'moment'
import { withRouter } from 'react-router-dom'
import validator from 'validator'

import { createVacation } from './scheduleActions'
import H1 from '../../components/H1'
import H3 from '../../components/H3'
import Day from './Day/index'
import Form from '../../components/Form'
import Icon from '../../components/Icon'
import Input from '../../components/FormInput'
import Checkbox from '../../components/FormCheckbox'

import './Schedule.css'

import type { RouterHistory } from 'react-router-dom'

type participant = {
  name: string,
  checked: boolean,
}

type Props = {
  auth: void | Object,
  date: Date,
  history: RouterHistory,
  createVacation: Function,
}

type State = {
  arrival: {
    value: string,
    touched: boolean,
  },
  departure: {
    value: string,
    touched: boolean,
  },
  people: Array<participant>,
  errors: {
    arrival: boolean | string,
    departure: boolean | string,
    people: boolean | string,
  },
}

export class Schedule extends Component<Props, State> {
  state = {
    arrival: {
      value: '',
      touched: false,
    },
    departure: {
      value: '',
      touched: false,
    },
    people: [
      {
        name: 'Folke',
        checked: false,
      },
      {
        name: 'Carina',
        checked: false,
      },
    ],
    errors: {
      arrival: '',
      departure: '',
      people: '',
    },
  }

  componentWillMount () {
    if (!this.props.auth) {
      this.props.history.push('/')
    }
  }

  renderIsWeekEnd (year: number, month: number, day: mixed) {
    if (!year || !month || typeof day !== 'number') return
    const weekDay = moment(`${year}/${month}/${day}`, 'YYYY/MM/D').format(
      'dddd'
    )
    return weekDay === 'Saturday' || weekDay === 'Sunday'
  }

  isToday (year: number, month: number, date: mixed) {
    if (!year || !month || typeof date !== 'number') return
    return (
      moment(`${year}-${month + 1}-${date}`, 'YYYY-MM-D').format(
        'YYYY-MM-DD'
      ) === moment().format('YYYY-MM-DD')
    )
  }

  renderDays () {
    const m = moment(this.props.date)
    const daysInMonth = m.daysInMonth()
    const year = m.year()
    const month = m.month()
    const startOnDayNr = moment(`${year}/${month}`, 'YYYY/MM').day()
    const days = []

    for (let i = 1; i < daysInMonth; i++) {
      days.push({
        date: i,
        occupied: false,
        status: '',
        weekend: this.renderIsWeekEnd(year, month, i),
        empty: false,
      })
    }

    for (let i = 1; i < startOnDayNr; i++) {
      days.unshift({
        date: null,
        occupied: false,
        status: '',
        weekend: false,
        empty: true,
      })
    }
    const emptyDaysAfter = 35 - days.length
    for (let i = 0; i < emptyDaysAfter; i++) {
      days.push({
        date: null,
        occupied: false,
        status: '',
        weekend: false,
        empty: true,
      })
    }

    return days.map(({ date, empty, occupied, status, weekend }, i) => (
      <Day
        date={date}
        empty={empty}
        key={`day-${i}`}
        occupied={occupied}
        status={status}
        today={this.isToday(year, month, date)}
        weekend={weekend}
      />
    ))
  }

  handleSubmit = (event: Event) => {
    event.preventDefault()
    this.validateForm()
    const people = []

    this.state.people.forEach(p => {
      if (p.checked === true) {
        people.push(p.name)
      }
    })

    this.props.createVacation({
      arrival: this.state.arrival.value,
      departure: this.state.departure.value,
      people,
    })
  }

  handleChange = (event: SyntheticInputEvent<any>) => {
    if (event && event.target && event.target.name) {
      const newState = { ...this.state }
      newState.people.map(p => {
        if (p.name === event.target.name) {
          return (p.checked = !p.checked)
        }
        return p
      })
      this.setState(newState)
    }
  }

  handleDateChange = (event?: SyntheticInputEvent<any>) => {
    if (event && event.target) {
      const target = event.target
      const value = target.type === 'checkbox' ? target.checked : target.value
      const name = target.name
      this.setState({ [name]: { ...this.state[name], value } })
    }
  }

  handleBlur = (field: string) => (event: SyntheticFocusEvent<any>) => {
    this.setState({
      [field]: { ...this.state[field], touched: true },
    })
    this.validateForm()
  }

  validateForm () {
    const errors = {}

    const arrival = this.state.arrival.value
    const departure = this.state.departure.value
    const people = this.state.people

    errors.arrival = !arrival.length && 'Arrival is required'
    if (!errors.arrival.length) {
      errors.arrival =
        !validator.isISO8601(arrival) && 'Arrival must be a valid Date.'
    }
    errors.departure = !departure.length && 'Departure is required'
    if (!errors.departure.length) {
      errors.departure =
        !validator.isISO8601(departure) && 'Departure must be a valid Date.'
    }
    errors.people = !people.length && 'At least one participant is required'

    this.setState({
      errors: { ...errors },
    })
  }

  renderParticipants () {
    const { people } = this.state
    return (
      people.length &&
      people.map((p, i) => (
        <Checkbox
          checked={p.checked}
          key={`participant-${i}`}
          label={p.name}
          name={p.name}
          onChange={this.handleChange}
        />
      ))
    )
  }

  renderCurrentMonth () {
    return moment(this.props.date).format('MMMM YYYY')
  }

  renderPreviousMonth () {
    const prevMonth = moment(this.props.date).clone()
    prevMonth.subtract(1, 'month')
    if (prevMonth.month() === 0 || prevMonth.month() === 11) {
      return prevMonth.format('MMMM YYYY')
    }
    return prevMonth.format('MMMM')
  }

  renderNextMonth () {
    const nextMonth = moment(this.props.date).clone()
    nextMonth.add(1, 'month').format('MMMM')
    if (nextMonth.month() === 0 || nextMonth.month() === 11) {
      return nextMonth.format('MMMM YYYY')
    }
    return nextMonth.format('MMMM')
  }

  showErrors = (field: string) => {
    if (this.state.errors[field] && this.state[field].touched) {
      return this.state.errors[field]
    }
  }

  render () {
    return (
      <section className="schedule">
        <H1>Schedule</H1>

        <nav className="time-nav">
          <Link to="/">
            <Icon marginRight name="arrow-circle-left" size="3x" />
            {this.renderPreviousMonth()}
          </Link>
          <span className="time-nav__current-month">
            {this.renderCurrentMonth()}
          </span>
          <Link to="/">
            {this.renderNextMonth()}
            <Icon marginLeft name="arrow-circle-right" size="3x" />
          </Link>
        </nav>
        <section className="calendar">
          <header className="calendar__header">
            <span>Monday</span>
            <span>Tuesday</span>
            <span>Wednesday</span>
            <span>Thursday</span>
            <span>Friday</span>
            <span>Saturday</span>
            <span>Sunday</span>
          </header>
          <main className="calendar__days">{this.renderDays()}</main>
        </section>

        <div className="schedule__inner">
          <Form onSubmit={this.handleSubmit} submitText="Add" title="New Trip">
            <H3>Time Period</H3>
            <div className="register__columns">
              <Input
                error={this.showErrors('arrival')}
                handleBlur={this.handleBlur('arrival')}
                handleChange={this.handleDateChange}
                label="Arrival"
                name="arrival"
                placeholder="YYYY-MM-DD"
                type="date"
                value={this.state.arrival.value}
              />
              <Input
                error={this.showErrors('departure')}
                handleBlur={this.handleBlur('departure')}
                handleChange={this.handleDateChange}
                label="Departure"
                name="departure"
                placeholder="YYYY-MM-DD"
                type="date"
                value={this.state.departure.value}
              />
            </div>
            <H3>Participants</H3>
            <div className="register__columns">{this.renderParticipants()}</div>
          </Form>
        </div>
      </section>
    )
  }
}

export const mapStateToProps = ({ auth, date }: Object) => {
  return {
    auth,
    date,
  }
}

export default connect(mapStateToProps, { createVacation })(
  withRouter(Schedule)
)
