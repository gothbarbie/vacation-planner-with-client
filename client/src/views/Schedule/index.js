import React from 'react'
import { Link } from 'react-router-dom'

import H1 from '../../components/H1'
import H3 from '../../components/H3'
import Day from './Day'
import Form from '../../components/Form'
import Icon from '../../components/Icon'
import Input from '../../components/Form/Input'

import './Schedule.css'

export default () => (
  <section className="schedule">
    <H1>January</H1>
    <nav className="time-nav">
      <Link to="/">
        <Icon 
          marginRight
          name="arrow-circle-left" 
          size="3x" />December 2017
      </Link>
      <Link to="/">
        February<Icon marginLeft 
          name="arrow-circle-right" 
          size="3x" /> 
      </Link>
    </nav>
    <section className="calendar">
      <header className="calendar__header">
        <span>Monday</span>
        <span>Tuesday</span>
        <span>Wednesday</span>
        <span>Thursday</span>
        <span>Friday</span>
        <span>Saturday</span>
        <span>Sunday</span>
      </header>
      <main className="calendar__days">
        <Day date={1} occupied status="Arrival" />
        <Day date={2} occupied status="Occupied" />
        <Day date={3} occupied status="Occupied" />
        <Day date={4} occupied status="Occupied" />
        <Day date={5} occupied status="Occupied" />
        <Day date={6} occupied status="Occupied" weekend />
        <Day date={7} occupied status="Departure" weekend />

        <Day date={8} />
        <Day date={9} />
        <Day date={10} />
        <Day date={11} />
        <Day date={12} />
        <Day date={13} weekend />
        <Day date={14} weekend />

        <Day date={15} />
        <Day date={16} />
        <Day date={17} />
        <Day date={18} />
        <Day date={19} />
        <Day date={20} weekend />
        <Day date={21} weekend />

        <Day date={22} />
        <Day date={23} />
        <Day date={24} />
        <Day date={25} />
        <Day date={26} />
        <Day date={27} weekend />
        <Day date={28} weekend />

        <Day date={29} />
        <Day date={30} />

        <Day empty />
        <Day empty />
        <Day empty />
        <Day empty />
        <Day empty />
      </main>
    </section>

    <Form>
      <H3>Content</H3>
      <Input label="Test" name="test" placeholder="placeholder" />
    </Form>
  </section>
)
