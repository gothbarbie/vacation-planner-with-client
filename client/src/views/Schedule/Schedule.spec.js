import React from 'react'
import { shallow } from 'enzyme'

import { Schedule } from './Schedule'

describe('views/Schedule', () => {
  const props = {
    auth: {},
    date: [],
  }
  const schedule = shallow(<Schedule {...props} />)

  it('renders', () => {
    expect(schedule).toMatchSnapshot()
  })
})
