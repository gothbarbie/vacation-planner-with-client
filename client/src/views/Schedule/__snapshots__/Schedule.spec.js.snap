// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`views/Schedule renders 1`] = `
<section
  className="schedule"
>
  <H1>
    Schedule
  </H1>
  <nav
    className="time-nav"
  >
    <Link
      replace={false}
      to="/"
    >
      <Icon
        marginRight={true}
        name="arrow-circle-left"
        size="3x"
      />
      February
    </Link>
    <span
      className="time-nav__current-month"
    >
      March 2018
    </span>
    <Link
      replace={false}
      to="/"
    >
      April
      <Icon
        marginLeft={true}
        name="arrow-circle-right"
        size="3x"
      />
    </Link>
  </nav>
  <section
    className="calendar"
  >
    <header
      className="calendar__header"
    >
      <span>
        Monday
      </span>
      <span>
        Tuesday
      </span>
      <span>
        Wednesday
      </span>
      <span>
        Thursday
      </span>
      <span>
        Friday
      </span>
      <span>
        Saturday
      </span>
      <span>
        Sunday
      </span>
    </header>
    <main
      className="calendar__days"
    >
      <Day
        date={null}
        empty={true}
        key="day-0"
        occupied={false}
        status=""
        weekend={false}
      />
      <Day
        date={null}
        empty={true}
        key="day-1"
        occupied={false}
        status=""
        weekend={false}
      />
      <Day
        date={null}
        empty={true}
        key="day-2"
        occupied={false}
        status=""
        weekend={false}
      />
      <Day
        date={1}
        empty={false}
        key="day-3"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={2}
        empty={false}
        key="day-4"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={3}
        empty={false}
        key="day-5"
        occupied={false}
        status=""
        today={false}
        weekend={true}
      />
      <Day
        date={4}
        empty={false}
        key="day-6"
        occupied={false}
        status=""
        today={false}
        weekend={true}
      />
      <Day
        date={5}
        empty={false}
        key="day-7"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={6}
        empty={false}
        key="day-8"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={7}
        empty={false}
        key="day-9"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={8}
        empty={false}
        key="day-10"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={9}
        empty={false}
        key="day-11"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={10}
        empty={false}
        key="day-12"
        occupied={false}
        status=""
        today={false}
        weekend={true}
      />
      <Day
        date={11}
        empty={false}
        key="day-13"
        occupied={false}
        status=""
        today={false}
        weekend={true}
      />
      <Day
        date={12}
        empty={false}
        key="day-14"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={13}
        empty={false}
        key="day-15"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={14}
        empty={false}
        key="day-16"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={15}
        empty={false}
        key="day-17"
        occupied={false}
        status=""
        today={true}
        weekend={false}
      />
      <Day
        date={16}
        empty={false}
        key="day-18"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={17}
        empty={false}
        key="day-19"
        occupied={false}
        status=""
        today={false}
        weekend={true}
      />
      <Day
        date={18}
        empty={false}
        key="day-20"
        occupied={false}
        status=""
        today={false}
        weekend={true}
      />
      <Day
        date={19}
        empty={false}
        key="day-21"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={20}
        empty={false}
        key="day-22"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={21}
        empty={false}
        key="day-23"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={22}
        empty={false}
        key="day-24"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={23}
        empty={false}
        key="day-25"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={24}
        empty={false}
        key="day-26"
        occupied={false}
        status=""
        today={false}
        weekend={true}
      />
      <Day
        date={25}
        empty={false}
        key="day-27"
        occupied={false}
        status=""
        today={false}
        weekend={true}
      />
      <Day
        date={26}
        empty={false}
        key="day-28"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={27}
        empty={false}
        key="day-29"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={28}
        empty={false}
        key="day-30"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={29}
        empty={false}
        key="day-31"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={30}
        empty={false}
        key="day-32"
        occupied={false}
        status=""
        today={false}
        weekend={false}
      />
      <Day
        date={null}
        empty={true}
        key="day-33"
        occupied={false}
        status=""
        weekend={false}
      />
      <Day
        date={null}
        empty={true}
        key="day-34"
        occupied={false}
        status=""
        weekend={false}
      />
    </main>
  </section>
  <div
    className="schedule__inner"
  >
    <Form
      onSubmit={[Function]}
      submitText="Add"
      title="New Trip"
    >
      <H3>
        Time Period
      </H3>
      <div
        className="register__columns"
      >
        <Input
          handleBlur={[Function]}
          handleChange={[Function]}
          label="Arrival"
          name="arrival"
          placeholder="YYYY-MM-DD"
          type="date"
          value=""
        />
        <Input
          handleBlur={[Function]}
          handleChange={[Function]}
          label="Departure"
          name="departure"
          placeholder="YYYY-MM-DD"
          type="date"
          value=""
        />
      </div>
      <H3>
        Participants
      </H3>
      <div
        className="register__columns"
      >
        <FormCheckbox
          checked={false}
          key="participant-0"
          label="Folke"
          name="Folke"
          onChange={[Function]}
        />
        <FormCheckbox
          checked={false}
          key="participant-1"
          label="Carina"
          name="Carina"
          onChange={[Function]}
        />
      </div>
    </Form>
  </div>
</section>
`;
